src/consensus/proposer.rs

```diff
@@ -297,56 +311,69 @@ impl BlockProposer {
             shard_id,
             address,
             proposed_blocks: BTreeMap::new(),
-            pending_chunks: BTreeMap::new(),
-            shard_decision_rx,
+            shard_stores,
             num_shards,
+            chain_id,
             block_tx,
             engine,
             statsd_client,
         }
     }
 
-    async fn collect_confirmed_shard_chunks(
+    async fn collect_confirmed_shard_witnesses(
         &mut self,
         height: Height,
         timeout: Duration,
-    ) -> Vec<ShardChunk> {
+    ) -> Vec<ShardChunkWitness> {
         let requested_height = height.block_number;
 
-        let mut poll_interval = time::interval(Duration::from_millis(10));
+        let mut poll_interval = time::interval(Duration::from_millis(100));
+        let mut chunks = BTreeMap::new();
 
         // convert to deadline
         let deadline = Instant::now() + timeout;
         loop {
             let timeout = time::sleep_until(deadline);
             select! {
                 _ = poll_interval.tick() => {
-                    // TODO(aditi): This breaks if syncd shard chunks show up in shard_decision_rx.
-                    if let Ok(chunk) = self.shard_decision_rx.try_recv() {
-                        let chunk_height = chunk.header.clone().unwrap().height.unwrap();
-                        let chunk_block_number = chunk_height.block_number;
-                        if self.pending_chunks.contains_key(&chunk_block_number) {
-                            self.pending_chunks.get_mut(&chunk_block_number).unwrap().push(chunk);
-                        } else {
-                            self.pending_chunks.insert(chunk_block_number, vec![chunk]);
+                    for (shard_id, store) in self.shard_stores.iter() {
+                        if chunks.contains_key(shard_id) {
+                            continue;
                         }
-                    }
-                    if let Some(chunks) = self.pending_chunks.get(&requested_height) {
-                        if chunks.len() == self.num_shards as usize {
-                            break;
+                        let result = store.shard_store.get_chunk_by_height(requested_height);
```

Comment:
Switches us to look in the db instead of waiting for the channel which was causing halts (if we lose a message, we could never recover).

We'll need to refactor once we want to allow block validators to run separately from shard validators, but that's going to take a while.